### PR TITLE
PE-930: validate names within folder

### PR DESCRIPTION
**Pull request by karlprieb**
This PR is based on #145 

- add `assertNamesWithinFolder` helper function and update unit tests
- add folder and file name validation to `createPublicFolderAndUploadChildren` and `createPrivateFolderAndUploadChildren` and create integration tests for both methods (just testing folder and name validation as they will be deprecated)

### CODE REVIEW

File: src/arfs/arfs_entity_name_validators.test.ts [hunk 1]
```diff
@@ -36,64 +40,157 @@ describe('entity name validators', () => {
...
+
+	type EntityTest = {
+		validationMethod: typeof assertValidArFSFileName;
+	} & Test;
+
+	type FolderWithChildrenTest = {
+		validationMethod: typeof assertNamesWithinFolder;
+	} & Test;
+
+	function isFolderWithChildren(test: EntityTest | FolderWithChildrenTest): test is FolderWithChildrenTest {
```
Comment (fedellen) [inline src/arfs/arfs_entity_name_validators.test.ts:57]: I don't think this type-guard is necessary since you've defined the `isFolderWithChildren` boolean on both types. Below, you could just use the boolean directly: `if (test.isFolderWithChildren)`

Also, I'm not sure if we're saving many lines of code here anymore by trying to fit this into the dynamic tests. Regardless, it the tests seem to be covering the flow just fine

**Review by fedellen — APPROVED** [review]
Looking good -- that will cover the names coming from those methods

Thanks for those integration tests!

File: src/arfs/arfs_entity_name_validators.test.ts (hunk 1 above)
Comment (karlprieb) [reply src/arfs/arfs_entity_name_validators.test.ts:57]: Yep, I tried to check for `test.isFolderWithChildren` and it didn't work as expected, that's why we have this type-guard.

We're not saving more lines, in fact we just added more lines 😄 But we're preventing test setup duplication and ensuring that we're running the same tests for all entity name validation functions.

Comment (fedellen) [reply src/arfs/arfs_entity_name_validators.test.ts:57]: Ah, I see how the compiler would be angry there

Comment (arielmelendez) [reply src/arfs/arfs_entity_name_validators.test.ts:57]: I find this test code hard to follow :( which thus makes assessing it's correctness more difficult to do. I much prefer test code duplication when I am forced to choose between readability/comprehensibility vs. elegance. The elegance doesn't help us if reviewers/code readers can't tell what the tests are doing without deep inspection.

File: src/arfs/arfs_entity_name_validators.ts [hunk 2]
```diff
@@ -38,3 +39,19 @@ export function assertValidArFSEntityNameFactory(entityType: EntityType): (name:
 		return true;
 	};
 }
+
+export function assertNamesWithinFolder(rootFolder: ArFSFolderToUpload, rootFolderDestName?: string): boolean {
+	assertValidArFSFolderName(rootFolderDestName ?? rootFolder.getBaseFileName());
```
Comment (arielmelendez) [inline src/arfs/arfs_entity_name_validators.ts:44]: It worries me that this logic (choosing the name) is duplicated here (it's in about 6 other places in the ArDrive class). It's a type design shortcoming with ArFSFolderToUpload. A built in function in that type that just tells me what the destination name is going to be would be ideal. Not an issue to fix here, but something that I want to broadcast.

File: src/arfs/arfs_entity_name_validators.ts [hunk 3]
```diff
@@ -38,3 +39,19 @@ export function assertValidArFSEntityNameFactory(entityType: EntityType): (name:
 		return true;
 	};
 }
+
+export function assertNamesWithinFolder(rootFolder: ArFSFolderToUpload, rootFolderDestName?: string): boolean {
```
Comment (arielmelendez) [inline src/arfs/arfs_entity_name_validators.ts:43]: Nit: could use a comment explaining what it does, though the name is pretty good :)

Comment (arielmelendez) [reply src/arfs/arfs_entity_name_validators.ts:43]: On second thought, the name could be improved: Consider `assertArFSCompliantNamesWithinFolder`

**Review by arielmelendez — APPROVED** [review]
Approved. Left some non-mission-critical feedback.
